src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -565,10 +564,36 @@ class G1RemSetSamplingTask : public G1ServiceTask {
     }
   }
 
+  // To avoid extensive rescheduling if the task is executed a bit early. The task is
+  // only rescheduled if the expected time is more than 1ms away.
+  bool should_reschedule() {
+    return reschedule_delay_ms() > 1;
+  }
+
+  // There is no reason to do the sampling if a GC occurred recently. We use the
+  // G1ConcRefinementServiceIntervalMillis as the metric for recently and calculate
+  // the diff to the last GC. If the last GC occurred longer ago than the interval
+  // 0 is returned.
+  jlong reschedule_delay_ms() {
+    Tickspan since_last_gc = G1CollectedHeap::heap()->time_since_last_collection();
+    jlong delay = (jlong) (G1ConcRefinementServiceIntervalMillis - since_last_gc.milliseconds());
+    return MAX2<jlong>(0L, delay);
+  }
+
 public:
   G1RemSetSamplingTask(const char* name) : G1ServiceTask(name) { }
   virtual void execute() {
-    sample_young_list_rs_length();
+    SuspendibleThreadSetJoiner sts;
+
+    // Reschedule if a GC happened too recently.
+    if (should_reschedule()) {
+      // Calculate the delay given the last GC and the interval.
+      schedule(reschedule_delay_ms());
+      return;
+    }
```

Comment:
I would prefer not having `should_reschedule` and reuse the return value of `reschedule_delay_ms`, but up to you.

```c
delay_ms = reschedule_delay_ms();
// only reschedule when ...
if (delay_ms > 1) {
  schedule(delay_ms);
  return;
}
```
Second point, why 1 in `reschedule_delay_ms() > 1`, but not 0?

Reply:
I tried to cover the why in the comment for `should_reschedule()`:
```
  // To avoid extensive rescheduling if the task is executed a bit early. The task is
  // only rescheduled if the expected time is more than 1ms away.
  bool should_reschedule() {
    return reschedule_delay_ms() > 1;
  }
```
But maybe this comment needs to be clearer. The problem is that when calculating the `reschedule_delay_ms()` the `since_last_gc.milliseconds()` call might round down to 299ms (when the interval is 300ms), even if we are just a microsecond away. And since there are different timestamps used to decide if the task is ready to run, we can end up in this situation.

Having this comment is one of the reasons I used two functions even if it meant calculating the delay two times. Makes sense? Would you like more info in the comment?

Reply:
> might round down to 299ms (when the interval is 300ms), even if we are just a microsecond away.

In that case, using `> 0` will just delay the start of the task for ~1ms, right? Doesn't sound like a serious problem. Similarly, `> 1` could cause the task to be run ~1ms earlier than its intended 300ms interval, right? Either way is fine, but I think `>0` is less surprising.

Reply:
One would think :) But the rounding comes back the other way around, if we reschedule to 1ms in the future, the service thread will look at the timestamp a bit later and then the time left will be rounded down to 0ms. So the effect will be that the task is rescheduled and run instantly over and over until the time since GC is over the interval.

Reply:
Resolved my confusion via private chat. Thank you for the explanation. This PR is good to go.